File: 42-trapping-rain-water/trapping-rain-water.go
/*

[0,1,0,2,1,0,1,3,2,1,2,1]
   0 
   i 
                     j
 lmax = 0
 rmax = 1
 water = min(lmax, rmax) - height => 

 if lmax < rmax: 
    can def store (lmax - height) water at i
    i++
    lmax: update
else:
    can def store (lmax - height) water at i
    j++
    rmax: update

*/

func trap(height []int) int {
    if len(height) <= 1 {
        return 0
    }
    lmax := height[0]
    rmax := height[len(height) - 1]

    i := 1
    j := len(height) - 2
    water := 0

    for i <= j {
        if lmax <= rmax{
            if lmax > height[i]{
                water += lmax - height[i]
            } else{
                lmax = height[i]
            }
            i++
        } else{
            if rmax > height[j]{
                water += rmax - height[j]
            } else {
                rmax = height[j]
            }
            j--
        }
    }
    return water
}

File: 42-trapping-rain-water/trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        """
        4 2 0 3 6 5
        0 2 4 1 0 0
                  l
                r
        lmax = 4
        rmax = 5
        each bar can only trap min(lmax, rmax) - itself


        """
        lmax, rmax = height[0], height[-1]
        l, r = 0, len(height) - 1
        water = 0

        while l <= r:
            if lmax <= rmax:
                water += max(lmax - height[l], 0)
                lmax = max(lmax, height[l])
                l += 1
            else:
                water += max(rmax - height[r], 0)
                rmax = max(rmax, height[r])
                r -= 1
        
        return water

